discv5/discv5-wire.md

```diff
@@ -49,187 +51,107 @@ request/response and 1s for the handshake.
+  cannot decrypt/authenticate the packet's message.
+- Handshake message packets, which are sent following WHOAREYOU. These packets establish a
+  new session and carry handshake-related data in addition to the encrypted/authenticated
+  message.
 
-Note that WHOAREYOU is only ever valid as a response to a previously sent request. If
-WHOAREYOU is received but no requests are pending, the handshake attempt can be ignored.
+In the following definitions, we assume that the sender of a packet has knowledge of its
+own 256-bit node ID (`src-id`) and the node ID of the packet destination (`dest-id`). When
+sending any packet except WHOAREYOU, the sender also generates a unique 96-bit `nonce`
+value.
 
-Implementations should be careful about AES-GCM nonces because encrypting two messages
-with the same nonce compromises the key. Session keys should be kept in memory for a
-limited amount of time, ensuring that nodes occasionally perform a handshake to establish
-new keys.
+### Protocol Header
 
-Implementations should also ensure that session secrets and the handshake are tied to a
-specific UDP endpoint. This is simple to implement by using the node ID and IP/port as the
-key into the in-memory session cache. When a node switches endpoints, e.g. when roaming
-between different wireless networks, sessions will to be re-established by re-handshaking.
-This requires no effort on behalf of the roaming node because the recipients of protocol
-messages will simply refuse to decrypt messages from the new endpoint and reply with
-WHOAREYOU.
+All discovery packets contain a header followed by an optional encrypted and authenticated
+message.
 
-**TBD: concurrent handshake tie-breaker rule.**
+Header information is 'masked' using symmetric encryption in order to avoid static
+identification of the protocol by firewalls.
 
-### Identity-Specific Cryptography in the Handshake
+    packet        = iv || masked-header || message
+    iv            = uint128   -- random data unique to packet
+    masked-header = aesctr_encrypt(masking-key, iv, header)
+    masking-key   = dest-id[:16]
 
-Establishment of session keys is dependent on the identity scheme of the recipient (i.e.
-the node which sends WHOAREYOU). Similarly, the signature over `id-nonce-input` is made by
-the identity key of the initiator. Although initiator and recipient might not be using the
-same identity scheme in their respective node records, implementations must be able to
-handle handshaking for all supported identity schemes.
+The `masked-header` contains the actual packet header, which starts with a fixed-size
+`static-header`, followed by a variable-length `authdata` section (of size `authdata-size`).
 
-At this time, the only supported identity scheme is "v4".
+    header        = static-header || authdata
+    static-header = protocol-id || src-id || flag || authdata-size
+    message       = aesgcm_encrypt(initiator-key, nonce, message-plaintext, header || authdata)
```

Comment:
```suggestion
    message       = aesgcm_encrypt(initiator-key, nonce, message-plaintext, header)
```
Is `header || authdata` the intention here?
I'd expect `header` or `static-header || authdata`

Reply:
I think `header` should be changed to `masked-header` and the `ad` part of the aesgcm_encrypt should be `masked-header`

Reply:
Additionally, the `authdata` can get rather big in case of a Handshake Message Packet. And while that should work, it seems unnecessary?

Reply:
You're right, the intention was `static-header || authdata`. I just forgot to update it when renaming the header.

The purpose of this authenticated data input is binding the message to the packet it is contained in. Processing of this data is very fast and I would prefer to put in as much of the packet as possible here.

Reply:
Thanks for the context on why. 

I understand you want to provide as much data as possible but I find it still weird because of the (big) difference of ad size between the two packet types, especially as you would be sending more often the `Ordinary Message Packet`. But ok, it does no harm I guess.